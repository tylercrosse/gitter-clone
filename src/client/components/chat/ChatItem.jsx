import React from 'react';
import moment from 'moment';

const ChatItem = ({ username, text, createdAt }) => {
  const dateString = moment(createdAt).format('MMM MM HH:mm');
  return (
    <article className="chat-item">
      <div className="chat-item-container">
        <div className="chat-item-aside">
          <img
            className="avatar"
            src={'http://i.pravatar.cc/30?u=' + username} alt={username}
          />
        </div>
        <div className="chat-item-content">
          <div className="chat-item-details">
            <div className="chat-item-user">{username}</div>
            <div className="chat-item-time">{dateString}</div>
          </div>
          <div className="chat-item-text">{text}</div>
        </div>
      </div>
    </article>
  );
};

export default ChatItem;
