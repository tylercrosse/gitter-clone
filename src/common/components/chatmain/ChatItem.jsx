import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import Avatar from '../Avatar';

export const ChatItem = ({
  User,
  text,
  rawMarkup,
  createdAt,
  burstStart
}) => {
  const dateString = moment(createdAt).format('MMM DD HH:mm');
  if (burstStart) {
    return (
      <article className="chat-item burst-start">
        <div className="chat-item-container">
          <div className="chat-item-aside">
            <Avatar user={User} />
          </div>
          <div className="chat-item-content">
            <div className="chat-item-details">
              <div className="chat-item-user">{User.name}</div>
              <div className="chat-item-time">{dateString}</div>
            </div>
            {rawMarkup ?
              <div
              className="chat-item-text"
              dangerouslySetInnerHTML={{__html: rawMarkup}}
              /> :
              <div className="chat-item-text">{text}</div>
            }
          </div>
        </div>
      </article>
    );
  }
  return (
    <article className="chat-item burst-continued">
      <div className="chat-item-container">
        <div className="chat-item-content">
          {rawMarkup ?
            <div
            className="chat-item-text"
            dangerouslySetInnerHTML={{__html: rawMarkup}}
            /> :
            <div className="chat-item-text">{text}</div>
          }
        </div>
      </div>
    </article>
  );
};

ChatItem.propTypes = {
  User: PropTypes.object.isRequired,
  text: PropTypes.string.isRequired,
  rawMarkup: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  burstStart: PropTypes.bool.isRequired,
};

export default ChatItem;
