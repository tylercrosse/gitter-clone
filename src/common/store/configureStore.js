import { createStore,
  applyMiddleware }             from 'redux';
import thunk                    from 'redux-thunk';
import { browserHistory }       from 'react-router';
import { autoRehydrate }        from 'redux-persist'
import { routerMiddleware }     from 'react-router-redux';
import { composeWithDevTools }  from 'redux-devtools-extension/developmentOnly';
import createSocketIoMiddleware from 'redux-socket.io';
import io                       from 'socket.io-client';
import { apiMiddleware }        from 'redux-api-middleware';
import rootReducer              from '../ducks/';

const socket = io('');
const socketIoMiddleware = createSocketIoMiddleware(socket, 'server.');

const middlewares = [
  thunk,
  apiMiddleware,
  socketIoMiddleware,
  routerMiddleware(browserHistory)
];

const configureStore = (preloadedState) => {
  const store = createStore(
    rootReducer,
    preloadedState,
    composeWithDevTools(
      applyMiddleware(...middlewares),
      autoRehydrate()
    )
  );

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept('../ducks/', () => {
      const nextRootReducer = require('../ducks/').default;
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};

export default configureStore;
