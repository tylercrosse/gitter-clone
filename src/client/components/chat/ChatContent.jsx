import React from 'react';
import { animateScroll } from 'react-scroll';
import burstify from '../../reducers/burst';
import ChatItem from './ChatItem.jsx';

class ChatContent extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      bursts: {}
    };
  }
  componentWillReceiveProps(nextProps) {
    this.setState({
      bursts: burstify(nextProps)
    });
  }
  componentDidUpdate() {
    animateScroll.scrollToBottom({containerId: 'chat-content'});
  }
  render() {
    return (
      <section
        id="chat-content"
        className="chat-content scroller"
      >
        {Object.values(this.props.messages).map((message) =>
          <ChatItem
            key={message._id}
            {...message}
          />
        )}
      </section>
    );
  }
}

export default ChatContent;
