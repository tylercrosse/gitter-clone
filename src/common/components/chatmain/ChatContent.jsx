import React, { PropTypes } from 'react';
import { animateScroll } from 'react-scroll';
import ChatItem from './ChatItem';

class ChatContent extends React.Component {
  componentDidUpdate() {
    animateScroll.scrollToBottom({containerId: 'chat-content'});
  }
  render() {
    const { messages } = this.props;
    return (
      <section
      id="chat-content"
      className="chat-content scroller"
      >
        {messages.length > 0 && messages.map((message) =>
          <ChatItem
          key={message.id}
          {...message}
          />
        )}
      </section>
    );
  }
}

ChatContent.propTypes = {
  messages: PropTypes.array.isRequired
};

export default ChatContent;
