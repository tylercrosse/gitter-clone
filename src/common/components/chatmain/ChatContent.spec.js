import React       from 'react';
import { Provider }   from 'react-redux';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { animateScroll }  from 'react-scroll';
import sinon       from 'sinon';
import ChatItem    from './ChatItem';
import ChatContent from './ChatContent';
import configureStore from '../../store/configureStore';

const setup = (propOverrides) => {
  const props = Object.assign({
    messages: [
      {
        id: '58911c3e871fdf24f2079782',
        createdAt: '2017-01-31T23:22:38.808Z',
        User: {name: 'liz'},
        text: 'sweet',
        rawMarkup: '<p>sweet</p>',
        burstStart: true
      }, {
        id: '58917370eb323841b839f583',
        createdAt: '2017-02-01T05:34:40.591Z',
        User: {name: 'dan'},
        text: 'cool',
        rawMarkup: '<p>cool</p>',
        burstStart: true
      }, {
        id: '5891737ceb323841b839f584',
        createdAt: '2017-02-01T05:34:52.011Z',
        User: {name: 'dan'},
        text: 'damn it',
        rawMarkup: '<p>damn it</p>',
        burstStart: false
      }
    ]
  }, propOverrides);
  const component = <ChatContent {...props} />;
  const wrapper = shallow(component);

  return {
    props,
    component,
    wrapper
  };
};

describe('<ChatContent />', () => {
  it('should render correctly', () => {
    const { wrapper } = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render a <ChatItem /> for each message', () => {
    const { wrapper } = setup();
    expect(wrapper.find(ChatItem))
      .toHaveLength(3);
  });

  describe('componentDidUpdate', () => {
    it('calls animateScroll.scrollToBottom() on update', () => {
      const { props } = setup();
      const mockStore = configureStore();
      const wrapper = mount(
        <Provider store={mockStore}>
          <ChatContent {...props} />
        </Provider>
      );
      const updateSpy = sinon.spy(ChatContent.prototype, 'componentDidUpdate');
      const scrollSpy = sinon.spy(animateScroll, 'scrollToBottom');
      wrapper.update();
      expect(updateSpy.calledOnce).toEqual(true);
      expect(scrollSpy.calledOnce).toEqual(true);
    });
  });
});
