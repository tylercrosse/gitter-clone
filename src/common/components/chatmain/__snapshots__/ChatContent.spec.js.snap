// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatContent /> should render correctly 1`] = `
<section
  className="chat-content scroller"
  id="chat-content"
>
  <Component
    User={
      Object {
        "name": "liz",
      }
    }
    burstStart={true}
    createdAt="2017-01-31T23:22:38.808Z"
    id="58911c3e871fdf24f2079782"
    rawMarkup="<p>sweet</p>"
    text="sweet"
  />
  <Component
    User={
      Object {
        "name": "dan",
      }
    }
    burstStart={true}
    createdAt="2017-02-01T05:34:40.591Z"
    id="58917370eb323841b839f583"
    rawMarkup="<p>cool</p>"
    text="cool"
  />
  <Component
    User={
      Object {
        "name": "dan",
      }
    }
    burstStart={false}
    createdAt="2017-02-01T05:34:52.011Z"
    id="5891737ceb323841b839f584"
    rawMarkup="<p>damn it</p>"
    text="damn it"
  />
</section>
`;
